P3379R0 Constrain std::expected equality operators

jensmaurer opened the pull request:
Fixes #7416 
Also fixes cplusplus/papers#2036

jwakely reviewed (changes requested):
There's a missing edit to [version.syn]:

> Update the value of the __cpp_lib_constrained_equality feature test macro in [version.syn].

jensmaurer commented:
@jwakely , fixed the missing version bump

I think we've decided before that feature-test macros are sort-of editorial, so I've added the `<expected>` amendment with a separate commit.
